Updating website & tutorials

tafsiri opened the pull request:
Making a PR to show the current progress and general direction I am taking. It will also make it easier to explain other proposals I have as I can rough things out here.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pair-code/deeplearnjs/673)
<!-- Reviewable:end -->

nsthorat commented:
Looking great!!

---

Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[README.md, line 72 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4q0-F3u6tej_PLCZ1a:-L4q0-F3u6tej_PLCZ1b:bxnb7ke) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/README.md#L72)):*
> ```Markdown
> var b = dl.scalar(2);
> 
> var result = dl.add(a, b);
> ```

use the chain api

---

*[README.md, line 75 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4q00BzePTPFRl-jnvO:-L4q00BzePTPFRl-jnvP:bwt4e3h) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/README.md#L75)):*
> ```Markdown
> 
> // Option 1: With a Promise.
> result.data().then((data) => console.log(data)); // Float32Array([3, 4, 5])
> ```

why add parens around data?

---

*[docs/tutorials/intro.md, line 96 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4q0CU3WOo0RSkK3LWu:-L4q0CU3WOo0RSkK3LWv:b57pd5i) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/docs/tutorials/intro.md#L96)):*
> ```Markdown
> 
> 
> #### WebGL backend
> ```

(my fault) Maybe we're going into too much implementation detail here and we can just explain how this manifests in the API? This will be the case regardless of backend

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-:-L4q0_wosOPEoRvadaKX:bpckfvv)*
<!-- Sent from Reviewable.io -->

dsmilkov commented:
Yup, this is awesome!

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[.gitignore, line 13 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4qeDZntRErM2GTGAV0:-L4qeDZntRErM2GTGAV1:b-5wrpu0) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/.gitignore#L13)):*
> ```
> .idea/
> deeplearn-*.tgz
> demos/homepage/vendor
> ```

run clean make-website. you'll probably won't need this?

---

*[README.md, line 53 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4qe_g55m7d4QvN0Cdg:-L4qe_g55m7d4QvN0Cdh:bbi7du6) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/README.md#L53)):*
> ```Markdown
> 
> const result = dl.add(a, b);
> result.data().then((data) => console.log(data)); // Float32Array([3, 4, 5]
> ```

just do `.then(data => ...)` (no parens)

---

*[README.md, line 55 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4qerM_COlWoumKSGCq:-L4qerM_COlWoumKSGCr:bpn3xsj) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/README.md#L55)):*
> final 

do we need to say `final`?

---

*[README.md, line 81 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4qf1cOLILN2aMGE63X:-L4qf1cOLILN2aMGE63Y:b-9jibc3) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/README.md#L81)):*
> ```Markdown
> ```
> 
> ## Development
> ```

would be great to move this section to DEVELOPMENT.md and maybe quickly link to it from the readme.md

---

*[demos/homepage/_config.yml, line 31 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4qfCXq7L0G6wDZZAVk:-L4qfCXq7L0G6wDZZAVl:b-pi9d9n) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/demos/homepage/_config.yml#L31)):*
> ```YAML
>   - demos/node_modules/
>   - demos/bower_components/
>   - vendor
> ```

remove this

---

*[docs/tutorials/intro.md, line 29 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4qjnhMM9Dpj8ApIpQM:-L4qjnhMM9Dpj8ApIpQN:b2v8en8) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/docs/tutorials/intro.md#L29)):*
> ES2017 syntax

just curious: are people generally ok with tutorials nowadays using es2017?

---

*[docs/tutorials/intro.md, line 65 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4qj5-wrhagRlDte-6l:-L4qj5-wrhagRlDte-6m:b-2hewd9) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/docs/tutorials/intro.md#L65)):*
> ```Markdown
> 
> ```js
> const a = dl.tensor2d([1.0, 2.0, 3.0, 4.0], [2, 2]);
> ```

how about `dl.tensor2d([[1, 2], [3, 4]])`, no need for `.0` suffix at each number, and no need for the shape. Here and elsewhere in this doc

---

*[docs/tutorials/intro.md, line 78 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4qjSr4EswROKbl7daI:-L4qjSr4EswROKbl7daJ:b-eec1s7) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/docs/tutorials/intro.md#L78)):*
> ```Markdown
>   const size = dl.scalar(a.size);
>   const average = dl.div(sum, size);
>   return average
> ```

missing semicolon. also chain api is more natural to read.

---

*[docs/tutorials/intro.md, line 96 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4q0CU3WOo0RSkK3LWu:-L4qkE0PS5I9Y0H0z0Zn:b-39ewb8) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/docs/tutorials/intro.md#L96)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

(my fault) Maybe we're going into too much implementation detail here and we can just explain how this manifests in the API? This will be the case regardless of backend
</blockquote></details>

+1. At least moving the backend details further down and bring training up

---

*[docs/tutorials/intro.md, line 206 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4qkQ9Cgf2QpnLrll02:-L4qkQ9Cgf2QpnLrll03:b-6dsicd) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/docs/tutorials/intro.md#L206)):*
> ```Markdown
> 
> ### Graphs
> **Note: the following sections describe using the deeplearn.js graph API. We have deprecated this API in support of a new 'eager' mode after research and community feedback. It will be removed in future versions of deeplearn.js. Eager Mode supports training and we will updated docs here soon.**
> ```

typo: we will *provide* updated docs here soon

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-:-L4qkVp7iRY0jCtmPJA6:b9x3kt1)*
<!-- Sent from Reviewable.io -->

tafsiri commented:
Review status: 3 of 5 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[README.md, line 55 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4qerM_COlWoumKSGCq:-L59eY-9j97gSjb6sIhD:b-dn9qk7) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/README.md#L55)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> final 

do we need to say `final`?
</blockquote></details>

Nope. Good catch.

---

*[README.md, line 72 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4q0-F3u6tej_PLCZ1a:-L59ekbNhe4eetvEHSyx:bdycq4e) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/README.md#L72)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

use the chain api
</blockquote></details>

updated. i also added a comment to indicate that a is not modified.

---

*[README.md, line 75 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4q00BzePTPFRl-jnvO:-L59fHCU9r3LelLg8aqM:b-hfpv0l) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/README.md#L75)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

why add parens around data?
</blockquote></details>

removed

---

*[README.md, line 81 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4qf1cOLILN2aMGE63X:-L59fJcE3hkEbM_0EZql:b-dwsh6j) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/README.md#L81)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

would be great to move this section to DEVELOPMENT.md and maybe quickly link to it from the readme.md
</blockquote></details>

Lets discuss this all together. I'm less of the opinion that README.md should be autoincluded here at all.  But maybe its worth it for now to make front page shorter?

---

*[docs/tutorials/intro.md, line 29 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4qjnhMM9Dpj8ApIpQM:-L59fhtaCx-wHXRXikPQ:b2f6wyk) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/docs/tutorials/intro.md#L29)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> ES2017 syntax

just curious: are people generally ok with tutorials nowadays using es2017?
</blockquote></details>

I think so. I see a lot of stuff written in ES2015+ nowadays. I was just trying to be more specific.

---

*[docs/tutorials/intro.md, line 65 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4qj5-wrhagRlDte-6l:-L59ftWFFCpMNWdV5XMu:biu6w0d) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/docs/tutorials/intro.md#L65)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

how about `dl.tensor2d([[1, 2], [3, 4]])`, no need for `.0` suffix at each number, and no need for the shape. Here and elsewhere in this doc
</blockquote></details>

I don't have super strong feelings about this, but i think its nice to introduce the concept of shape (and being able to specify it even when providing a flat array.  Maybe it would be nice if `b` below was created with an inferred shape. I'll do that for now so we can see how it looks.

I also added a shape inference example to the tensors section

---

*[docs/tutorials/intro.md, line 96 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4q0CU3WOo0RSkK3LWu:-L59ph3Uk-YGn7YoEhHL:bru3awr) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/docs/tutorials/intro.md#L96)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

+1. At least moving the backend details further down and bring training up
</blockquote></details>

backends have been moved to bottom (just before graph)

---

*[docs/tutorials/intro.md, line 206 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4qkQ9Cgf2QpnLrll02:-L59plRkdHm-gsTS3IeU:b-78jop0) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/docs/tutorials/intro.md#L206)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

typo: we will *provide* updated docs here soon
</blockquote></details>

I actually remove the we will provide updated docs soon as we have the new Training for ML Beginners

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673)*
<!-- Sent from Reviewable.io -->

tafsiri commented:
@nsthorat @dsmilkov this is ready for closer review now.

nsthorat commented:
Review status: 1 of 8 files reviewed at latest revision, 54 unresolved discussions, some commit checks failed.

---

*[.gitignore, line 13 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4qeDZntRErM2GTGAV0:-L5EUnqz8O3pgU1x5_pc:bvu98yf) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/.gitignore#L13)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

run clean make-website. you'll probably won't need this?
</blockquote></details>

Reminder to do this

---

*[README.md, line 12 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EUs3GH76U_KAg-DMO:-L5EUs3GH76U_KAg-DMP:bb7p4x8) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/README.md#L12)):*
> ```Markdown
> or run pre-trained models in inference mode.
> 
> We provide an API the closely mirrors the TensorFlow eager API.
> ```

that

---

*[README.md, line 26 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EVgLYQmpltCtU8WRW:-L5EVgLYQmpltCtU8WRX:b-4nnuxp) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/README.md#L26)):*
> ```Markdown
> 
> ```html
> <script src="https://cdn.jsdelivr.net/npm/deeplearn"></script>
> ```

why not combine these blocks, add @latest here and then say you can replace latest with @version? Otherwise this block is really long and scary

---

*[README.md, line 51 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EVstPrm1QrvdBqc8e:-L5EVstPrm1QrvdBqc8f:bn6dn7i) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/README.md#L51)):*
> ```Markdown
> const b = dl.scalar(2);
> 
> const result = a.add(b); // a is not modified, result is a new tensor/
> ```

trailing slash

---

*[README.md, line 52 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EVwarcOgvWV080dy9:-L5EVwarcOgvWV080dyA:b-5kn27s) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/README.md#L52)):*
> ```Markdown
> 
> const result = a.add(b); // a is not modified, result is a new tensor/
> result.data().then(data => console.log(data)); // Float32Array([3, 4, 5]
> ```

I just added a "print" method, maybe that will be better to use here and below. In the introduction we can explain data and dataSync to get the values 

`a.add(b).print();`

---

*[README.md, line 71 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EWHEz3v9E3NmvgFtk:-L5EWHEz3v9E3NmvgFtl:bn6dn7i) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/README.md#L71)):*
> ```Markdown
> var b = dl.scalar(2);
> 
> var result = a.add(b); // a is not modified, result is a new tensor/
> ```

trailing slash

---

*[demos/homepage/index.md, line 82 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EWWELyYdQl1cTKK_G:-L5EWWELyYdQl1cTKK_H:bb7p4x8) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/demos/homepage/index.md#L82)):*
> ```Markdown
>   <div class="mdl-layout-spacer mdl-cell--hide-tablet mdl-cell--hide-phone"></div>
>   <div class= "intro-text mdl-cell mdl-cell--4-col mdl-cell--8-col-tablet mdl-cell-4-col-phone">
>     <p class='intro-body mdl-typography--body-1'>We provide an API the closely mirrors
> ```

that

---

*[docs/tutorials/intro.md, line 65 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4qj5-wrhagRlDte-6l:-L5EWyAKC4PDUzAFdBG8:b-9mv5h4) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/docs/tutorials/intro.md#L65)):*
<details><summary><i>Previously, tafsiri (Yannick Assogba) wrote…</i></summary><blockquote>

I don't have super strong feelings about this, but i think its nice to introduce the concept of shape (and being able to specify it even when providing a flat array.  Maybe it would be nice if `b` below was created with an inferred shape. I'll do that for now so we can see how it looks.

I also added a shape inference example to the tensors section
</blockquote></details>

I think it's good to show examples with shapes passed explicitly

---

*[docs/tutorials/intro.md, line 78 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4qjSr4EswROKbl7daI:-L5EX0H8jnmwEDaGGJj9:b-bk0tpg) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/docs/tutorials/intro.md#L78)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

missing semicolon. also chain api is more natural to read.
</blockquote></details>

Maybe we can show `print` and then another line saying if you want the values back as an array do this stuff

---

*[docs/tutorials/intro.md, line 12 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EWaPLZn19iNTQCUN8:-L5EWaPLZn19iNTQCUN9:b-ofj73d) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L12)):*
> ```Markdown
> in a browser or run pre-trained models in inference mode.
> 
> Lets take a look at some of the core concepts in deeplean.js
> ```

deeplean

---

*[docs/tutorials/intro.md, line 17 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EWdD8tsQqqT2v8GXs:-L5EWdD8tsQqqT2v8GXt:bz9pwz) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L17)):*
> ```Markdown
> 
> The central unit of data in **deeplearn.js** is the `Tensor`. A `Tensor`
> consists of a set of floating point values shaped into an array of an arbitrary
> ```

no longer just floating point, maybe just vlaues

---

*[docs/tutorials/intro.md, line 28 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EWrmULyZhqD0py_Q3:-L5EWrmULyZhqD0py_Q4:bh1n91u) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L28)):*
> ```Markdown
> let a = dl.tensor2d([1.0, 2.0, 3.0, 10.0, 20.0, 30.0], shape);
> // deeplearn.js can also infer the shape
> let b = dl.tensor2d([[0.0, 2.0], [4.0, 6.0]]);
> ```

maybe do shape inference first

---

*[docs/tutorials/intro.md, line 53 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EXBcGY5hniVXdAuQ4:-L5EXBcGY5hniVXdAuQ5:b-t8y86k) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L53)):*
> ```Markdown
> _upload_ that tensor to the GPU automatically. When we want to use it in out JavaScript context
> (e.g. to print it out), we call `data()` or `dataSync()` to _download_ it to the CPU memory. Note that
> this is a relatively expensive operation, so you would likely want to call the async version.
> ```

I don't understand this last part - why would you want to call the sync version if this is expensive?

---

*[docs/tutorials/intro.md, line 58 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EXI164ZQFDDPdVfx4:-L5EXI164ZQFDDPdVfx5:b-54kz1k) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L58)):*
> ```Markdown
> ### Operations (Ops)
> 
> While Tensors allow us to store data, ops allow us to manipulate data. Deeplean.js comes with a wide variety of mathematical opearations suitable for linear algebra and machine learning. These include unary ops like `square()` and binary ops like `add()` and `mul()` Generally speaking an op will do some transformation on one of more tensors and return a new tensor as a result.
> ```

use ticks for Tensor

---

*[docs/tutorials/intro.md, line 58 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EXKCnSzDAskpZFONB:-L5EXKCnSzDAskpZFONC:bwsrcui) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L58)):*
> ```Markdown
> ### Operations (Ops)
> 
> While Tensors allow us to store data, ops allow us to manipulate data. Deeplean.js comes with a wide variety of mathematical opearations suitable for linear algebra and machine learning. These include unary ops like `square()` and binary ops like `add()` and `mul()` Generally speaking an op will do some transformation on one of more tensors and return a new tensor as a result.
> ```

`deeplearn.js` lowercase, and with an r :)

---

*[docs/tutorials/intro.md, line 61 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EXRzca5WdLH3yeIed:-L5EXRzca5WdLH3yeIee:b-lx76az) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L61)):*
> ```Markdown
> 
> ```js
> let a = dl.tensor2d([1.0, 2.0, 3.0, 4.0], [2, 2]);
> ```

I think we should do shape inference here and for the rest of the doc for simplicity

---

*[docs/tutorials/intro.md, line 76 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EX_KQ0TB-zlW49UKN:-L5EX_KQ0TB-zlW49UKO:b-2lsvim) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L76)):*
> ```Markdown
> 
> Because deeplearn.js uses the GPU to accelerate math operations, there is a need
> to manage GPU memory. While in regular javascript this is handled with scopes, we
> ```

JavaScript

---

*[docs/tutorials/intro.md, line 109 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EZwuPzGTdzgE4jX7i:-L5EZwuPzGTdzgE4jX7j:bcx8lwd) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L109)):*
> ```Markdown
> But using tidy functions is much more convenient!
> 
> ## Training
> ```

maybe somewhere here say this is being computed eagerly?

---

*[docs/tutorials/intro.md, line 111 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EXmLirnmGTBXtWXMj:-L5EXmLirnmGTBXtWXMk:b9rcbxs) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L111)):*
> ```Markdown
> ## Training
> 
> At the heart of many machine learning problems is the question of actually _training_ the machine to do some task. In deeplearn.js this process is encapsulated by  _Optimizers_. Optimizers are strategies to progressively tune the variables of your model in order to reduce the error (or _loss_ in ML parlance) in your models predictions.
> ```

model's or models'

---

*[docs/tutorials/intro.md, line 119 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EXrq5ht0mmol1ran6:-L5EXrq5ht0mmol1ran7:bh9kumn) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L119)):*
> ```Markdown
> 
> // Initialize the models variables
> const weights = dl.variable(dl.randNormal([10, 64]));
> ```

randomNormal

---

*[docs/tutorials/intro.md, line 124 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EXtmHB8q5CluVdOlE:-L5EXtmHB8q5CluVdOlF:benpa5j) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L124)):*
> ```Markdown
> // Set a learning rate and create an optimizer.
> const LEARNING_RATE = .1;
> const optimizer = new dl.SGDOptimizer(LEARNING_RATE);
> ```

dl.train.sgd(LEARNING_RATE)

---

*[docs/tutorials/intro.md, line 127 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EY1jfWKkz2faqiXw0:-L5EY1jfWKkz2faqiXw1:b48dr1k) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L127)):*
> ```Markdown
> 
> /**
>  * Perform inference
> ```

maybe the jsdoc and inline comments are redundant, maybe keep one or the other?

---

*[docs/tutorials/intro.md, line 134 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EXy8MAeH_OyJeG7Hj:-L5EXy8MAeH_OyJeG7Hk:bpqlrxb) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L134)):*
> ```Markdown
> 
> /**
>  * Compute the loss of the model by comparing the prediction
> ```

Maybe here you can say something about how this is supposed to return a scalar value, which is important.

---

*[docs/tutorials/intro.md, line 137 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EYCyVCdZdSHGQW9a-:-L5EYCyVCdZdSHGQW9a0:bga4xjo) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L137)):*
> ```Markdown
>  * and ground truth.
>  */
> function loss(predictions, groundTruth) {
> ```

can you call groundTruth label just to bring it closer to what people say in the literature more often?

---

*[docs/tutorials/intro.md, line 170 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EYVclZAyNtsRLRX7T:-L5EYNQNr4UA1-wgVZHA:b-w099xb) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L170)):*
> ```Markdown
>  * Train the model.
>  */
> function train(data, batchSize, numEpochs) {
> ```

since this wont change too much between standard models, maybe we can show it?

Also, maybe remove batchSize since data will own its batchSize (outer dimension)

---

*[docs/tutorials/intro.md, line 171 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EYfG47UzylHnGhxou:-L5EYfG47UzylHnGhxov:b-o6lhel) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L171)):*
> ```Markdown
>  */
> function train(data, batchSize, numEpochs) {
>   // Call trainstep in a loop.  Use requestAnimationFrame to
> ```

`await dl.nextFrame()`-- I'm about to export this

---

*[docs/tutorials/intro.md, line 176 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5E_2Ln9mAPlspm2tk_:-L5E_2Ln9mAPlspm2tka:b22xuxn) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L176)):*
> ```Markdown
> ```
> 
> ## Backends
> ```

I wonder if this should just go in another separate md file, maybe "backends". after looking through this, this page is quite a mouthful. what do you think?

---

*[docs/tutorials/intro.md, line 178 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EZeAPXEperlNqgzjq:-L5EZeAPXEperlNqgzjr:b-w3mxtt) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L178)):*
> ```Markdown
> ## Backends
> 
> The library provides a number of _backends_ which implement the core operations of the library, currently we have a __CPU__ backend and a __WebGL__ backend. Deeplearn.js will use the __WebGL__ backend by default whenever the browser supports it. The __WebGL__ backend uses the computers' __GPU__, to perform fast and highly optimized linear algebra operations.
> ```

s/operations/kernels

s/Deeplearn.js/deeplearn.js

---

*[docs/tutorials/intro.md, line 204 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5E_EHGZPzcf6ifRB-t:-L5E_EHGZPzcf6ifRB-u:b-h9wb1a) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L204)):*
> ```Markdown
> const diff = dl.sub(a, b);
> 
> // Calling .data returns a promise that resolves to a TypedArray that holds
> ```

consider `.print()`

---

*[docs/tutorials/ml_beginners.md, line 7 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EaDB8xNlQ7Oiqq6Pz:-L5EaDB8xNlQ7Oiqq6Q-:b-oyz6us) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/ml_beginners.md#L7)):*
> ```Markdown
> # Training for ML Beginners
> 
> In order to get a model to learn to do the things we would like, we have to train it. There are various styles of training, but this guide will primarily focus on supervised training. This is where we build a model and give it some examples of inputs and __desired outputs__, we then adjust our model variables to learn to correctly 'map' from these inputs to outputs and hopefully generalize to unseen inputs.
> ```

in the second sentence, replace "training" with "learning"

and maybe say  __desired outputs__, aka "labels"

---

*[docs/tutorials/ml_beginners.md, line 9 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EaNyJu3L3ep2em6Ip:-L5EaNyJu3L3ep2em6Iq:b-vn363h) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/ml_beginners.md#L9)):*
> ```Markdown
> In order to get a model to learn to do the things we would like, we have to train it. There are various styles of training, but this guide will primarily focus on supervised training. This is where we build a model and give it some examples of inputs and __desired outputs__, we then adjust our model variables to learn to correctly 'map' from these inputs to outputs and hopefully generalize to unseen inputs.
> 
> Lets look at a toy example to get a feel for what training a model looks like. Note that this tutorials _does not_ include machine learning best practices that deeplearn.js supports such as batching or shuffling training examples.
> ```

tutorial

---

*[docs/tutorials/ml_beginners.md, line 17 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EaRqwZ3kfQ7YxK8MJ:-L5EaRqxx9ilYHQqtlSA:bu1nkjb) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/ml_beginners.md#L17)):*
> ```Markdown
> ```
> 
> To do this the key deeplearn.js API we will use is the *optimizer* API. We will set up some variables (or weights) to represent the coefficients (a, b and c), and the optimizer will automatically adjust the values of these coefficients. Lets take a look at the code, it is heavily commented so feel free to read through it slowly.
> ```

How about simply: `To do this we will use the optimizer API.`

---

*[docs/tutorials/ml_beginners.md, line 17 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5Ea_dm_wQrnlFHmQ-w:-L5Ea_dm_wQrnlFHmQ-x:b-84ydmm) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/ml_beginners.md#L17)):*
> ```Markdown
> ```
> 
> To do this the key deeplearn.js API we will use is the *optimizer* API. We will set up some variables (or weights) to represent the coefficients (a, b and c), and the optimizer will automatically adjust the values of these coefficients. Lets take a look at the code, it is heavily commented so feel free to read through it slowly.
> ```

Let's

---

*[docs/tutorials/ml_beginners.md, line 39 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EbOekvccmOAqohfrg:-L5EbOekvccmOAqohfrh:bdcivmj) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/ml_beginners.md#L39)):*
> ```Markdown
> // Step 1. Set up variables, these are the things we want the model
> // to learn in order to do prediction accurately. We will initialize
> // them with random variables.
> ```

s/variables/values

---

*[docs/tutorials/ml_beginners.md, line 47 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EbPyuVlowZE_6Tkfq:-L5EbPyuVlowZE_6Tkfr:bspk8ab) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/ml_beginners.md#L47)):*
> ```Markdown
> // Step 2. Create an optimizer, we will use this later
> const learningRate = 0.01;
> const optimizer = new dl.SGDOptimizer(learningRate);
> ```

`const optimizer = dl.train.sgd(learningRate)`

---

*[docs/tutorials/ml_beginners.md, line 59 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EbdDEWFsF7u4ionQV:-L5EbdDEWFsF7u4ionQW:bq4o7m3) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/ml_beginners.md#L59)):*
> ```Markdown
>  * directly model the quadratic equation we are trying to model.
>  *
>  * It is also sometimes referred to as the "forward" step of our training process.
> ```

single quotes for consistency

---

*[docs/tutorials/ml_beginners.md, line 81 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EbiHYcqqHQhQ7HQ6y:-L5EbiHYcqqHQhQ7HQ6z:b-p0hig2) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/ml_beginners.md#L81)):*
> ```Markdown
>  * This will tell us how good a prediction is from what we expect.
>  *
>  * prediction is a tensor with our predicted y value
> ```

quotes around prediction

---

*[docs/tutorials/ml_beginners.md, line 82 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5Ebk0m-tZQXN8jetG7:-L5Ebk0m-tZQXN8jetG8:b-m4n3ei) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/ml_beginners.md#L82)):*
> ```Markdown
>  *
>  * prediction is a tensor with our predicted y value
>  * actual number is a number with the y value the model should have predicted
> ```

period

---

*[docs/tutorials/ml_beginners.md, line 102 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EbnBb_KIJ1cInDEFn:-L5EbnBb_KIJ1cInDEFo:bcz2g4g) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/ml_beginners.md#L102)):*
> ```Markdown
> 
>   // We use requestAnimationFrame to not block the browser ui.
>   requestAnimationFrame(function trainStep() {
> ```

use dl.nextFrame so that this looks like a standard for loop that you would see in other ML examples.

```
for (let i = 0; i < xs.length; i++) {
  minimize();
  await dl.nextFrame();
}
```

---

*[docs/tutorials/ml_beginners.md, line 106 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5Ec--1aHPh5wFGkCyt:-L5Ec--1aHPh5wFGkCyu:b-vknwlu) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/ml_beginners.md#L106)):*
> ```Markdown
>       // Minimize is where the magic happens, we must return a
>       // numerical estimate of how well we are doing using the
>       // current state of the variables we created at the start.
> ```

e.g. Loss

---

*[docs/tutorials/ml_beginners.md, line 109 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EcDpa3GqjQLCokyIf:-L5EcDpa3GqjQLCokyIg:bm08qgj) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/ml_beginners.md#L109)):*
> ```Markdown
> 
>       // This optimizer does the 'backward' step of our training data
>       // updating variables defined previously based on the amount error
> ```

instead of saying based on the amount (since this isn't always true), how about say something like `to minimize the loss`

---

*[docs/tutorials/ml_beginners.md, line 149 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EcSEMNOqNITcOsVi7:-L5EcSENcvBVHatf8YpS:b6rf2wa) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/ml_beginners.md#L149)):*
> ```Markdown
> 
> // Lets see how it does before training.
> console.log("Before training: using random coefficients")
> ```

single quotes

---

*[docs/tutorials/ml_beginners.md, line 157 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EcVKvIpJL2_hMYfZc:-L5EcVKvIpJL2_hMYfZd:b-d2pl3p) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/ml_beginners.md#L157)):*
> ```Markdown
> });
> 
> // Huzzah We have trained a simple machine learning model!
> ```

lower case W

---

*[docs/tutorials/ml_beginners.md, line 164 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EckE18rjWHdTTUqwU:-L5EckE18rjWHdTTUqwV:b-mccq7u) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/ml_beginners.md#L164)):*
> ```Markdown
> value for "y" given an "x".
> 
> This is a very simple model and training setup. Too simple for most challenges we would want to use neural networks for. In practice we would likely not just use `Scalar` variables. We would likely want to use techniques like batching our training data, which allows us to run prediction over a number of inputs in parallel, and computing error over a batch. As well as shuffling our training data, so that the model doesn't learn something dependent on the order of examples. Future tutorials will explore common techniques used in building more robust training pipelines.
> ```

when you say "instead of Scalar", maybe say 

In practice, we would likely not just use `Scalar` variables, but higher-dimensional `Tensors`, including images.

Or something to that effect

---

*[docs/tutorials/ml_beginners.md, line 164 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EcvI-b8uoCE-mY_8-:-L5EcvI-b8uoCE-mY_80:bouefom) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/ml_beginners.md#L164)):*
> ```Markdown
> value for "y" given an "x".
> 
> This is a very simple model and training setup. Too simple for most challenges we would want to use neural networks for. In practice we would likely not just use `Scalar` variables. We would likely want to use techniques like batching our training data, which allows us to run prediction over a number of inputs in parallel, and computing error over a batch. As well as shuffling our training data, so that the model doesn't learn something dependent on the order of examples. Future tutorials will explore common techniques used in building more robust training pipelines.
> ```

Sentence that starts with `As well as shuffling` is not a complete sentence.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673)*
<!-- Sent from Reviewable.io -->

tafsiri commented:
Review status: 0 of 7 files reviewed at latest revision, 54 unresolved discussions, some commit checks failed.

---

*[README.md, line 53 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4qe_g55m7d4QvN0Cdg:-L5B-IbIsmh6vArB18yz:bvnrpcm) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/README.md#L53)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

just do `.then(data => ...)` (no parens)
</blockquote></details>

updated

---

*[README.md, line 26 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EVgLYQmpltCtU8WRW:-L5EkSyvRsW30NgVeMmh:bjhb47d) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/README.md#L26)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

why not combine these blocks, add @latest here and then say you can replace latest with @version? Otherwise this block is really long and scary
</blockquote></details>

Good suggestion. How would you feel about dropping one of the CDN's as well? :)

---

*[README.md, line 52 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EVwarcOgvWV080dy9:-L5El1KvJNmoH9YGVeYx:b-9jf44z) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/README.md#L52)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

I just added a "print" method, maybe that will be better to use here and below. In the introduction we can explain data and dataSync to get the values 

`a.add(b).print();`
</blockquote></details>

So i think in this case we want to show data(), it just happens that console is the easiest thing to do.

---

*[docs/tutorials/intro.md, line 65 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4qj5-wrhagRlDte-6l:-L5EmpHTKlNQaQ-UnaQv:bwqzpjb) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/docs/tutorials/intro.md#L65)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

I think it's good to show examples with shapes passed explicitly
</blockquote></details>

So current state is one tensor has explicit shape and the other has inferred shape.

---

*[docs/tutorials/intro.md, line 78 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4qjSr4EswROKbl7daI:-L5Emvq9ld0STUvY-NZB:b-pz9r26) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/docs/tutorials/intro.md#L78)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

Maybe we can show `print` and then another line saying if you want the values back as an array do this stuff
</blockquote></details>

Like above i think the main point here is to show data(), console.log just being convenient. Also i think this is natural thing to want to see (the values). 

---

*[docs/tutorials/intro.md, line 17 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EWdD8tsQqqT2v8GXs:-L5EmYv7R3OEwKz_1LQz:b3e5cen) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L17)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

no longer just floating point, maybe just vlaues
</blockquote></details>

how about numerical?

---

*[docs/tutorials/intro.md, line 28 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EWrmULyZhqD0py_Q3:-L5Emf10tELsaNoaeQpJ:b-ld6e7y) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L28)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

maybe do shape inference first
</blockquote></details>

I think in this order it makes it clearer what 'shape' is. I.e. 2 rows 3 columns.

---

*[docs/tutorials/intro.md, line 53 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EXBcGY5hniVXdAuQ4:-L5En9VqD1jkBXJQBhC4:bq5nlcz) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L53)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

I don't understand this last part - why would you want to call the sync version if this is expensive?
</blockquote></details>

Not sure i understand. I think this suggests calling the 'async' version. Let me know if its not clear and i'll rewrite it anyway to make to clearer.

---

*[docs/tutorials/intro.md, line 58 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EXI164ZQFDDPdVfx4:-L5EnTiR5CAzbbJn-DQR:b-5u0vnt) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L58)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

use ticks for Tensor
</blockquote></details>

so i think it looks awkward for the plural form `Tensor`s looks awkward to me when rendered.

---

*[docs/tutorials/intro.md, line 109 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EZwuPzGTdzgE4jX7i:-L5EoMsTTKRhQ-5c1uct:bl13odq) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L109)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

maybe somewhere here say this is being computed eagerly?
</blockquote></details>

added something, take a look

---

*[docs/tutorials/intro.md, line 127 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EY1jfWKkz2faqiXw0:-L5EpIVgm_2Q3zxRFzvN:b-1ice69) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L127)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

maybe the jsdoc and inline comments are redundant, maybe keep one or the other?
</blockquote></details>

kept the jsdoc.

---

*[docs/tutorials/intro.md, line 137 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EYCyVCdZdSHGQW9a-:-L5EpPyt_2iiODoYfigk:b-rd3c9d) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L137)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

can you call groundTruth label just to bring it closer to what people say in the literature more often?
</blockquote></details>

sure. i think its worse though. label is fairly ambiguous. 

---

*[docs/tutorials/intro.md, line 170 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EYVclZAyNtsRLRX7T:-L5EpfvK3fwxUhRE9MeF:b6tquvf) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L170)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

since this wont change too much between standard models, maybe we can show it?

Also, maybe remove batchSize since data will own its batchSize (outer dimension)
</blockquote></details>

I kinda thought of leaving it to the training tutorials as it can get long and we can show some good practices/starter patterns there.

I'm not sure what the last sentence means. I think a user can organize batchsize however they want?

---

*[docs/tutorials/intro.md, line 176 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5E_2Ln9mAPlspm2tk_:-L5EqJAFfXohRo7K05ZW:b22y7vo) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L176)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

I wonder if this should just go in another separate md file, maybe "backends". after looking through this, this page is quite a mouthful. what do you think?
</blockquote></details>

I think its okay here (and kinda fits in context imo).

---

*[docs/tutorials/intro.md, line 178 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EZeAPXEperlNqgzjq:-L5EqWehGlIlr4qkhfzT:bmyvnl3) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L178)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

s/operations/kernels

s/Deeplearn.js/deeplearn.js
</blockquote></details>

Does kernels mean something to someone not intimately familiar with machine learning implementations? I'm guessing the desire is to separate higher level operations from lower level kernels, but is there a more generic term we can use? How about "core building blocks" in the first past. "Linear algebra operations" still sounds right though (over linear algebra kernels). Thoughts?

---

*[docs/tutorials/intro.md, line 204 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5E_EHGZPzcf6ifRB-t:-L5EqzLj3dll7HdUShgB:bumhuih) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L204)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

consider `.print()`
</blockquote></details>

Again i think its more about the downloading the data. Also print is blocking, do we want to generally encourage that?

---

*[docs/tutorials/ml_beginners.md, line 9 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EaNyJu3L3ep2em6Ip:-L5EriTaYP44K100ynOK:b-qpco2l) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/ml_beginners.md#L9)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

tutorial
</blockquote></details>

Ping @dsmilkov, does this sentence satisfy what you wanted to get across about the things we are not showing in this particular example?

---

*[.gitignore, line 13 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L4qeDZntRErM2GTGAV0:-L5B-HBsAX59ILHYOVaO:b-intrgr) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5ee290549a479ee7bf62b9c88ec0e3467d1ba416/.gitignore#L13)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

run clean make-website. you'll probably won't need this?
</blockquote></details>

worked

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673)*
<!-- Sent from Reviewable.io -->

nsthorat commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Review status: 0 of 7 files reviewed at latest revision, 19 unresolved discussions.

---

*[README.md, line 26 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EVgLYQmpltCtU8WRW:-L5EwcZIl1Gq3u4KE3TR:b-tugtct) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/README.md#L26)):*
<details><summary><i>Previously, tafsiri (Yannick Assogba) wrote…</i></summary><blockquote>

Good suggestion. How would you feel about dropping one of the CDN's as well? :)
</blockquote></details>

Sure, drop one

---

*[docs/tutorials/intro.md, line 28 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EWrmULyZhqD0py_Q3:-L5ExDfQnCGhJqgSyrbo:b-ff2wb3) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L28)):*
<details><summary><i>Previously, tafsiri (Yannick Assogba) wrote…</i></summary><blockquote>

I think in this order it makes it clearer what 'shape' is. I.e. 2 rows 3 columns.
</blockquote></details>

show b's shape

---

*[docs/tutorials/intro.md, line 109 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EZwuPzGTdzgE4jX7i:-L5Exz5a_olbevUQAqWG:b-9fmci8) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L109)):*
<details><summary><i>Previously, tafsiri (Yannick Assogba) wrote…</i></summary><blockquote>

added something, take a look
</blockquote></details>

sorry, my bad, let's roll this back

---

*[docs/tutorials/intro.md, line 137 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EYCyVCdZdSHGQW9a-:-L5EyGm7A9O6seHLNdUG:b-2x2tr9) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L137)):*
<details><summary><i>Previously, tafsiri (Yannick Assogba) wrote…</i></summary><blockquote>

sure. i think its worse though. label is fairly ambiguous. 
</blockquote></details>

s/label/labels

---

*[docs/tutorials/intro.md, line 170 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EYVclZAyNtsRLRX7T:-L5EylSVtgjk48sbr7am:bxekfwr) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L170)):*
<details><summary><i>Previously, tafsiri (Yannick Assogba) wrote…</i></summary><blockquote>

I kinda thought of leaving it to the training tutorials as it can get long and we can show some good practices/starter patterns there.

I'm not sure what the last sentence means. I think a user can organize batchsize however they want?
</blockquote></details>

change batchsize to a comment

---

*[docs/tutorials/intro.md, line 178 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EZeAPXEperlNqgzjq:-L5Ez3jQzc1LITFpObRb:b-c873mj) ([raw file](https://github.com/pair-code/deeplearnjs/blob/8aaca2ffad6133685fc28f89857bc5a685644317/docs/tutorials/intro.md#L178)):*
<details><summary><i>Previously, tafsiri (Yannick Assogba) wrote…</i></summary><blockquote>

Does kernels mean something to someone not intimately familiar with machine learning implementations? I'm guessing the desire is to separate higher level operations from lower level kernels, but is there a more generic term we can use? How about "core building blocks" in the first past. "Linear algebra operations" still sounds right though (over linear algebra kernels). Thoughts?
</blockquote></details>

operations => mathematics

operations => kernels

---

*[docs/tutorials/intro.md, line 144 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-L5EyjvAcqh2723_P7eT:-L5EyjvAcqh2723_P7eU:b-o3b264) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5b225e345d8915ab34e62a9d1a5d6d22c466f3aa/docs/tutorials/intro.md#L144)):*
> ```Markdown
>  * Train the model a *single* step.
>  */
> function trainStep(data, label, returnCost = true) {
> ```

labels

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/673#-:-L5F-e6PhNJ2Nj-QgTo6:b-jbsf8z)*
<!-- Sent from Reviewable.io -->
